### PR TITLE
Add sanitizers

### PR DESCRIPTION
**Pull request by Spartan322**
Adds UndefinedBehaviorSanitizer, AddressSanitizer, LeakSanitizer, ThreadSanitizer, and MemorySanitizer to Linux builds
Adds AddressSanitizer to Windows builds
Adds UndefinedBehaviorSanitizer, AddressSanitizer, and ThreadSanitizer to Mac builds

### CODE REVIEW

File: tools/macos.py [hunk 1]
```diff
@@ -12,6 +13,9 @@ def options(opts):
     opts.Add("macos_sdk_path", "macOS SDK path", "")
     if has_osxcross():
         opts.Add("osxcross_sdk", "OSXCross SDK version", "darwin16")
+    opts.Add(BoolVariable("use_ubsan", "Use LLVM/GCC compiler undefined behavior sanitizer (UBSAN)", False))
```
Comment (wvpm) [inline tools/macos.py:16]: Does macos not support msan and lsan?

File: tools/windows.py [hunk 2]
```diff
@@ -11,6 +11,7 @@ def options(opts):
     opts.Add(BoolVariable("use_mingw", "Use the MinGW compiler instead of MSVC - only effective on Windows", False))
     opts.Add(BoolVariable("use_clang_cl", "Use the clang driver instead of MSVC - only effective on Windows", False))
     opts.Add(BoolVariable("use_static_cpp", "Link MinGW/MSVC C++ runtime libraries statically", False))
+    opts.Add(BoolVariable("use_asan", "Use address sanitizer (ASAN)", False))
```
Comment (wvpm) [inline tools/windows.py:14]: Why no tsan and others for windows?

**Review by wvpm — APPROVED** [review]

File: tools/macos.py (hunk 1 above)
Comment (Spartan322) [reply tools/macos.py:16]: Yeah not with native compilation tools. google/sanitizers#1026

File: tools/windows.py (hunk 2 above)
Comment (Spartan322) [reply tools/windows.py:14]: MSVC doesn't support any other sanitizer.

**Pull request merged**